lib/db.js

```diff
@@ -0,0 +1,381 @@
+        value: [output.satoshis, script, height].join(':')
+      });
+    }
+
+    if(tx.isCoinbase()) {
+      continue;
+    }
+
+  }
+
+  setImmediate(function() {
+    callback(null, operations);
+  });
+};
+
+DB.prototype._onChainAddBlock = function(block, callback) {
+
+  var self = this;
+
+  log.debug('DB handling new chain block');
+
+  // Remove block from mempool
+  self.mempool.removeBlock(block.hash);
+
+  async.series([
+    this._updateOutputs.bind(this, block, true), // add outputs
+  ], function(err, results) {
+
+    if (err) {
+      return callback(err);
+    }
+
+    var operations = [];
+    for (var i = 0; i < results.length; i++) {
+      operations = operations.concat(results[i]);
+    }
+
+    log.debug('Updating the database with operations', operations);
+
+    self.store.batch(operations, callback);
+
+  });
+
+};
+
+
+DB.prototype._onChainRemoveBlock = function(block, callback) {
+
+  var self = this;
+
+  async.series([
+    this._updateOutputs.bind(this, block, false), // remove outputs
+  ], function(err, results) {
+
+    if (err) {
+      return callback(err);
+    }
+
+    var operations = [];
+    for (var i = 0; i < results.length; i++) {
+      operations = operations.concat(results[i]);
+    }
+    self.store.batch(operations, callback);
+
+  });
+
+};
+
+DB.prototype.getAPIMethods = function() {
+  return [
+    ['getTransaction', this, this.getTransaction, 2],
+    ['getBalance', this, this.getBalance, 2],
+    ['getOutputs', this, this.getOutputs, 2],
+    ['getUnspentOutputs', this, this.getUnspentOutputs, 2],
+    ['isSpent', this, this.isSpent, 2]
+  ];
+};
+
+DB.prototype.getBalance = function(address, queryMempool, callback) {
+  this.getUnspentOutputs(address, queryMempool, function(err, outputs) {
+    if(err) {
+      return callback(err);
+    }
+
+    var satoshis = outputs.map(function(output) {
+      return output.satoshis;
+    });
+
+    var sum = satoshis.reduce(function(a, b) {
+      return a + b;
+    }, 0);
+
+    return callback(null, sum);
+  });
+};
+
+DB.prototype.getOutputs = function(address, queryMempool, callback) {
+  var self = this;
+
+  var outputs = [];
+  var key = [DB.PREFIXES.OUTPUTS, address].join('-');
+
+  var stream = this.store.createReadStream({
+    start: key,
+    end: key + '~'
+  });
+
+  stream.on('data', function(data) {
+
+    var key = data.key.split('-');
+    var value = data.value.split(':');
+
+    var output = {
+      address: key[1],
+      txid: key[3],
+      outputIndex: Number(key[4]),
+      satoshis: Number(value[0]),
+      script: value[1],
+      blockHeight: Number(value[2])
+    };
+
+    outputs.push(output);
+
+  });
+
+  var error;
+
+  stream.on('error', function(streamError) {
+    if (streamError) {
+      error = streamError;
+    }
+  });
+
+  stream.on('close', function() {
+    if (error) {
+      return callback(error);
+    }
+
+    /*if(queryMempool) {
+      var mempoolOutputs = self._getMempoolOutputs(address);
```

Comment:
`daemon.getMempoolOutputs()` is ready to be added in PR: https://github.com/bitpay/bitcoind.js/pull/33